st2auth_ldap/ldap_backend.py

```diff
@@ -155,6 +155,12 @@ def authenticate(self, username, password):
             # Instantiate connection object and bind with service account.
             try:
                 connection = self._init_connection()
+
+                if self._bind_dn.find('{username}') != -1:
+                    self._bind_dn = self._bind_dn.format(username=username)
+                if self._bind_password.find('{password}') != -1:
```

Comment:
While there's nothing inherently wrong with using the supplied username/password to bind and do the check since if the user account is valid and can be authenticated, it should be able to query the group membership. This also addresses policies if faceless/service account is not allowed. However, instead of reusing bind_dn and bind_password and doing a string replacement, the code should be clear that these are optional now and if not given in the config file, the backend will use the supplied username/password.  We shouldn't reuse a variable (which has different meaning or format) just for the sake of making this an easy change. Also, where're the unit and integration tests here?

Reply:
Some additional check is required here to make sure username is not empty string to avoid an anonymous bind with LDAP.